# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources (README.md file)
- Happy learning!

In [1]:
# import numpy and pandas
import pandas as pd
import numpy as np
from scipy.stats import ttest_1samp
import scipy.stats as st
import matplotlib.pyplot as plt

# Challenge 1 - Exploring the Data

In this challenge, we will examine all salaries of employees of the City of Chicago. We will start by loading the dataset and examining its contents.

In [2]:
chicago = pd.read_csv("Current_Employee_Names__Salaries__and_Position_Titles.csv")


FileNotFoundError: [Errno 2] No such file or directory: 'Current_Employee_Names__Salaries__and_Position_Titles.csv'

Examine the `salaries` dataset using the `head` function below.

In [ ]:
chicago.head()



Name                              Job Titles  \
0    AARON,  JEFFERY M                                SERGEANT   
1      AARON,  KARINA   POLICE OFFICER (ASSIGNED AS DETECTIVE)   
2  AARON,  KIMBERLEI R                CHIEF CONTRACT EXPEDITER   
3  ABAD JR,  VICENTE M                       CIVIL ENGINEER IV   
4    ABASCAL,  REECE E             TRAFFIC CONTROL AIDE-HOURLY   

         Department Full or Part-Time Salary or Hourly  Typical Hours  \
0            POLICE                 F           Salary            NaN   
1            POLICE                 F           Salary            NaN   
2  GENERAL SERVICES                 F           Salary            NaN   
3       WATER MGMNT                 F           Salary            NaN   
4              OEMC                 P           Hourly           20.0   

   Annual Salary  Hourly Rate  
0       101442.0          NaN  
1        94122.0          NaN  
2       101592.0          NaN  
3       110064.0          NaN  
4            NaN        19.86

We see from looking at the `head` function that there is quite a bit of missing data. Let's examine how much missing data is in each column. Produce this output in the cell below

In [ ]:
chicago_null = chicago.isnull().sum()
chicago_null



Name                     0
Job Titles               0
Department               0
Full or Part-Time        0
Salary or Hourly         0
Typical Hours        25161
Annual Salary         8022
Hourly Rate          25161
dtype: int64

Let's also look at the count of hourly vs. salaried employees. Write the code in the cell below

In [ ]:
chicago['Salary or Hourly'].value_counts()


Salary    25161
Hourly     8022
Name: Salary or Hourly, dtype: int64

What this information indicates is that the table contains information about two types of employees - salaried and hourly. Some columns apply only to one type of employee while other columns only apply to another kind. This is why there are so many missing values. Therefore, we will not do anything to handle the missing values.

There are different departments in the city. List all departments and the count of employees in each department.

In [ ]:
# Your code here:
chicago.groupby("Department").agg({"Name": "count"})

Name
Department                  
ADMIN HEARNG              39
ANIMAL CONTRL             81
AVIATION                1629
BOARD OF ELECTION        107
BOARD OF ETHICS            8
BUDGET & MGMT             46
BUILDINGS                269
BUSINESS AFFAIRS         171
CITY CLERK                84
CITY COUNCIL             411
COMMUNITY DEVELOPMENT    207
COPA                     116
CULTURAL AFFAIRS          65
DISABILITIES              28
DoIT                      99
FAMILY & SUPPORT         615
FINANCE                  560
FIRE                    4641
GENERAL SERVICES         980
HEALTH                   488
HUMAN RELATIONS           16
HUMAN RESOURCES           79
INSPECTOR GEN             87
LAW                      407
LICENSE APPL COMM          1
MAYOR'S OFFICE            85
OEMC                    2102
POLICE                 13414
POLICE BOARD               2
PROCUREMENT               92
PUBLIC LIBRARY          1015
STREETS & SAN           2198
TRANSPORTN              1140
TREASURER                 22
WATER MGMNT             1879

# Challenge 2 - Hypothesis Tests

In this section of the lab, we will test whether the hourly wage of all hourly workers is significantly different from $30/hr. Import the correct one sample test function from scipy and perform the hypothesis test for a 95% two sided confidence interval.

In [ ]:
# Your code here:
#H0 = The average hourly wage of hourly workers is equal to $30/hour.
#H1 = The average hourly wage of hourly workers is different from $30/hour.

# 2. Significance level
alpha = 0.05
mu = 30

# 3. Sample
sample = chicago[chicago["Salary or Hourly"]== "Hourly"]["Hourly Rate"]

# 4. Compute statistics / 5. Get p-value
t_statistic, p_value = st.ttest_1samp(sample, mu)
print(st.ttest_1samp(sample, mu))

if p_value < alpha:
    print("Reject the null hypothesis: The average hourly wage is significantly different from $30/hr.")
else:
    print("Fail to reject the null hypothesis: The average hourly wage is not significantly different from $30/hr.")

TtestResult(statistic=20.6198057854942, pvalue=4.3230240486229894e-92, df=8021)
Reject the null hypothesis: The average hourly wage is significantly different from $30/hr.


We are also curious about salaries in the police force. The chief of police in Chicago claimed in a press briefing that salaries this year are higher than last year's mean of $86000/year a year for all salaried employees. Test this one sided hypothesis using a 95% confidence interval.

Hint: A one tailed test has a p-value that is half of the two tailed p-value. If our hypothesis is greater than, then to reject, the test statistic must also be positive.

In [ ]:
#Hypothesis
#H0 = salaries >= $86000/year for all salaried employees
#H1 =  salaries < $86000

#Significance
alpha = 0.05

#Sample
sample = chicago[chicago["Salary or Hourly"] == "Salary"]
mu = 86000
annual_salary = sample["Annual Salary"]

#Compute stastic
t_statistic, p_value = st.ttest_1samp(annual_salary, mu, alternative='less')

print(f"T-statistic: {t_statistic}")
print(f"P-value: {p_value}")

if p_value < alpha:
    print("We can reject the null hypothesis")
else:
    print("We can not reject the null hypothesis")

T-statistic: 5.932870515690814
P-value: 0.9999999984921207
We can not reject the null hypothesis


Using the `crosstab` function, find the department that has the most hourly workers. 

In [ ]:
# Your code here:

cross_tab = pd.crosstab(chicago['Department'], chicago['Salary or Hourly'])

department_with_most_hourly_workers = cross_tab['Hourly'].idxmax()

hourly_workers_in_streets_san = cross_tab.loc['STREETS & SAN', 'Hourly']

print("Department with the most hourly workers:", department_with_most_hourly_workers)
print("Number of hourly workers in STREETS & SAN:", hourly_workers_in_streets_san)


Department with the most hourly workers: STREETS & SAN
Number of hourly workers in STREETS & SAN: 1862


The workers from the department with the most hourly workers have complained that their hourly wage is less than $35/hour. Using a one sample t-test, test this one-sided hypothesis at the 95% confidence level.

In [ ]:
# Your code here:
# Hypothesis 
#H0 = Workers from Streets & San hourly wage <= 35$
# H1 = Workers from Streets & San hourly wage > 35$

#Significance
alpha = 0.05

#Sample
#on my sample i choose 560 from my population 1862 Streets & San (about 30% of the hourly workers)
sample = chicago[(chicago["Department"]=="STREETS & SAN") & (chicago["Salary or Hourly"] =="Hourly")]["Hourly Rate"].sample(560)
mu = 35

#stats
t_statistic, p_value = st.ttest_1samp(sample, mu, alternative="greater")

print(f"T-statistic: {t_statistic}")
print(f"P-value: {p_value}")

if p_value < alpha:
    print("We can reject the null hypothesis")
else:
    print("We cannot reject the null hypothesis")

T-statistic: -5.096278326234201
P-value: 0.9999997626711307
We cannot reject the null hypothesis


# Challenge 3: To practice - Constructing Confidence Intervals

While testing our hypothesis is a great way to gather empirical evidence for accepting or rejecting the hypothesis, another way to gather evidence is by creating a confidence interval. A confidence interval gives us information about the true mean of the population. So for a 95% confidence interval, we are 95% sure that the mean of the population is within the confidence interval. 
).

To read more about confidence intervals, click [here](https://en.wikipedia.org/wiki/Confidence_interval).


In the cell below, we will construct a 95% confidence interval for the mean hourly wage of all hourly workers. 

The confidence interval is computed in SciPy using the `t.interval` function. You can read more about this function [here](https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.stats.t.html).

To compute the confidence interval of the hourly wage, use the 0.95 for the confidence level, number of rows - 1 for degrees of freedom, the mean of the sample for the location parameter and the standard error for the scale. The standard error can be computed using [this](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.sem.html) function in SciPy.

In [ ]:
# Your code here:

#sample
hourly_workers = chicago[chicago['Salary or Hourly'] == 'Hourly']

#Mean and std error
mean_hourly_wage = hourly_workers['Hourly Rate'].mean()
standard_error = st.sem(hourly_workers['Hourly Rate'])

#confidence level
confidence_level = 0.95

#degrees of freedom
degrees_of_freedom = len(hourly_workers) - 1

#confidence interval using t.interval
confidence_interval = st.t.interval(confidence_level, degrees_of_freedom, loc=mean_hourly_wage, scale=standard_error)

print(f"95% Confidence Interval for Mean Hourly Wage: {confidence_interval}")

95% Confidence Interval for Mean Hourly Wage: (32.52345834488425, 33.05365708767623)


***My insights***
I'm  quite confident that the true average hourly wage of all hourly workers lies between approximately $32.52 and $33.05 with a 95% confidence level. This provides me an idea of the likely range within which the true average lies based on the available sample.

Now construct the 95% confidence interval for all salaried employeed in the police in the cell below.

In [ ]:
#sample
salaried_employeed = chicago[(chicago["Department"]=="POLICE") & (chicago["Salary or Hourly"] =="Salary")]["Annual Salary"]

#confidance level
confidance_level = 0.95
alpha = 1 - confidance_level

#degrees of freedom
ddof = len(salaried_employeed) - 1

#compute mean and std error
mean = salaried_employeed.mean()
standard_error = st.sem(salaried_employeed)

# Calculate the confidence interval using t.interval
confidence_interval = st.t.interval(alpha, df=ddof, loc=mean, scale=standard_error)

print(f"95% Confidence Interval for Annual Salary employees in Police: {confidence_interval}")

95% Confidence Interval for Annual Salary employees in Police: (86476.5176546444, 86496.31135162238)


***My insights***
I'm quite confident that the true average annual salary of employees in the "POLICE" department lies between approximately $86476.52 and $86496.31 with a 95% confidence level. 

# Bonus Challenge - Hypothesis Tests of Proportions

Another type of one sample test is a hypothesis test of proportions. In this test, we examine whether the proportion of a group in our sample is significantly different than a fraction. 

You can read more about one sample proportion tests [here](http://sphweb.bumc.bu.edu/otlt/MPH-Modules/BS/SAS/SAS6-CategoricalData/SAS6-CategoricalData2.html).

In the cell below, use the `proportions_ztest` function from `statsmodels` to perform a hypothesis test that will determine whether the number of hourly workers in the City of Chicago is significantly different from 25% at the 95% confidence level.

In [ ]:
# Your code here:

